pathology/hovernet/training_ignite.py

```diff
@@ -0,0 +1,303 @@
+    labels = np.load(os.path.join(data_dir, "labels.npy"))
+
+    data = [
+        {
+            "image": image,
+            "image_meta_dict": {"original_channel_dim": -1},
+            "label": label,
+            "label_meta_dict": {"original_channel_dim": -1},
+        }
+        for image, label in zip(images, labels)
+    ]
+
+    train_data = [data[i] for i in train_indices]
+    valid_data = [data[i] for i in valid_indices]
+
+    return train_data, valid_data
+
+def cropping_center(x, crop_shape, batch=False):
+    """Crop an input image at the centre.
+    Args:
+        x: input array
+        crop_shape: dimensions of cropped array
+
+    Returns:
+        x: cropped array
+
+    """
+    orig_shape = x.shape
+    if not batch:
+        h0 = int((orig_shape[1] - crop_shape[0]) * 0.5)
+        w0 = int((orig_shape[2] - crop_shape[1]) * 0.5)
+        x = x[:, h0: h0 + crop_shape[0], w0: w0 + crop_shape[1]]
+    else:
+        h0 = int((orig_shape[2] - crop_shape[0]) * 0.5)
+        w0 = int((orig_shape[3] - crop_shape[1]) * 0.5)
+        x = x[..., h0: h0 + crop_shape[0], w0: w0 + crop_shape[1]]
+    return x
+
+def val_post_process(data):
+    seg_pred = data["pred"][HoVerNetBranch.NP.value]
+    trans = Compose([Activations(softmax=True), AsDiscrete(argmax=True, to_onehot=2)])
+    data["pred"][HoVerNetBranch.NP.value] = trans(seg_pred)
+
+    return data
+
+def train_post_process(data):
+    seg_pred = data["pred"][HoVerNetBranch.NP.value]
+    trans = Compose([Activations(softmax=True), AsDiscrete(argmax=True, to_onehot=2)])
+    data["pred"][HoVerNetBranch.NP.value] = trans(seg_pred)
+
+    return data
+
+def _from_engine(keys):
+    keys = ensure_tuple(keys)
+
+    def _wrapper(data):
+        ret = [[i[k][HoVerNetBranch.NP.value] for i in data] for k in keys]
+        return tuple(ret) if len(ret) > 1 else ret[0]
+
+    return _wrapper
+
+
+class PrepareBatchExtraInput_v2():
+    def __init__(self, extra_keys) -> None:
+        self.prepare_batch = PrepareBatchExtraInput(extra_keys)
+
+    def __call__(self, batchdata, device, non_blocking, **kwargs):
+        image, label, extra_label, _ = self.prepare_batch(batchdata, device, non_blocking, **kwargs)
+        all_label = {
+            HoVerNetBranch.NP: label,
+            HoVerNetBranch.NC: extra_label[0],
+            HoVerNetBranch.HV: extra_label[1],
+        }
+
+        return image, all_label
+
+
+def run(cfg):
+    log_dir = create_log_dir(cfg)
+    device = set_device(cfg)
+    # --------------------------------------------------------------------------
+    # Data Loading and Preprocessing
+    # --------------------------------------------------------------------------
+    # __________________________________________________________________________
+    # __________________________________________________________________________
+    # Build MONAI preprocessing
+    train_data, valid_data = prepare_datasets(cfg["root"])
+    train_transforms = Compose(
```

Comment:
Do we need any random transforms?
Refer to: https://github.com/Dana-Farber-AIOS/pathml/blob/master/examples/train_hovernet.ipynb

Thanks.

Reply:
Yes! Have added it, thanks!